File: src/safety.py
"""Safety primitives: input validation, repo cleanup, and advisory locking.

Addresses: C3 (repo name injection), C5 (shell=True prohibition),
           H5 (ensure_clean invariant), M4 (flock concurrency guard).
"""
import fcntl
import re
import subprocess
from contextlib import contextmanager
from pathlib import Path
from typing import Generator

_REPO_NAME_RE = re.compile(r'^[A-Za-z0-9._-]+$')
_RESERVED_NAMES = frozenset({".", ".."})

# CR-MED-4: strict github URL match — no extra path, no query, no whitespace.
_HTTPS_URL_RE = re.compile(r'^https://github\.com/[A-Za-z0-9._-]+/[A-Za-z0-9._-]+(\.git)?$')
_SSH_URL_RE = re.compile(r'^git@github\.com:[A-Za-z0-9._-]+/[A-Za-z0-9._-]+(\.git)?$')


def validate_repo_name(name: str) -> None:
    """Raise ValueError if *name* is not a safe, single-component repository name.

    Accepted: strings matching ``^[A-Za-z0-9._-]+$`` that are not ``.`` or ``..``.
    Rejected: empty strings, path separators, shell metacharacters, reserved names.
    """
    if not name:
        raise ValueError(f"repo name must not be empty")
    if name in _RESERVED_NAMES:
        raise ValueError(f"unsafe repo name (reserved): {name!r}")
    if not _REPO_NAME_RE.match(name):
        raise ValueError(f"unsafe repo name (invalid characters): {name!r}")


def validate_ssh_url(url: str) -> None:
    """Raise ValueError if *url* is not an allowed GitHub clone URL.

    Accepted:
    - ``git@github.com:<owner>/<repo>[.git]``
    - ``https://github.com/<owner>/<repo>``

    All other schemes, hosts, or injected flags are rejected.
    """
    if not url:
        raise ValueError("clone URL must not be empty")
    if _HTTPS_URL_RE.match(url) or _SSH_URL_RE.match(url):
        return
    raise ValueError(f"unexpected clone URL: {url!r}")


def ensure_clean(repo_dir: Path) -> None:
    """Reset *repo_dir* to a clean state, removing all local modifications.

    Steps:
    1. ``git reset --hard HEAD`` — restores tracked files to HEAD state.
    2. ``git clean -fd`` — removes untracked files and directories.
    3. Deletes ``MERGE_HEAD``, ``CHERRY_PICK_HEAD``, ``REBASE_HEAD`` if present,
       so git does not consider the repo mid-operation.

    All subprocess calls use list form with ``shell=False``.
    """
    subprocess.run(
        ["git", "reset", "--hard", "HEAD"],
        cwd=repo_dir,
        check=False,
        capture_output=True,
    )
    subprocess.run(
        ["git", "clean", "-fd"],
        cwd=repo_dir,
        check=False,
        capture_output=True,
    )
    for marker in ("MERGE_HEAD", "CHERRY_PICK_HEAD", "REBASE_HEAD"):
        (repo_dir / ".git" / marker).unlink(missing_ok=True)


@contextmanager
def acquire_lock(path: Path) -> Generator[None, None, None]:
    """Advisory exclusive lock on *path* using ``fcntl.flock``.

    Raises ``BlockingIOError`` immediately if another process holds the lock
    (``LOCK_NB`` — non-blocking). The lock file is created if it does not exist.
    The lock is released when the context exits, even if an exception is raised.

    Usage::

        with acquire_lock(Path("/run/myapp/lock")):
            do_work()
    """
    path.parent.mkdir(parents=True, exist_ok=True)
    fd = open(path, "w")
    try:
        fcntl.flock(fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
        try:
            yield
        finally:
            fcntl.flock(fd, fcntl.LOCK_UN)
    finally:
        fd.close()

File: go/internal/safety/safety.go
// Package safety: input validation, repo cleanup, advisory locking, blast-radius.
package safety

import (
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var (
	repoNameRe  = regexp.MustCompile(`^[A-Za-z0-9._-]+$`)
	httpsURLRe  = regexp.MustCompile(`^https://github\.com/[A-Za-z0-9._-]+/[A-Za-z0-9._-]+(\.git)?$`)
	sshURLRe    = regexp.MustCompile(`^git@github\.com:[A-Za-z0-9._-]+/[A-Za-z0-9._-]+(\.git)?$`)
	reservedSet = map[string]bool{".": true, "..": true}
)

// ErrLocked indicates the lock is already held by another process.
var ErrLocked = errors.New("lock held by another process")

// ValidateRepoName rejects empty, ".", "..", and any name with shell/path metacharacters.
func ValidateRepoName(name string) error {
	if name == "" {
		return errors.New("repo name must not be empty")
	}
	if reservedSet[name] {
		return fmt.Errorf("unsafe repo name (reserved): %q", name)
	}
	if !repoNameRe.MatchString(name) {
		return fmt.Errorf("unsafe repo name (invalid characters): %q", name)
	}
	return nil
}

// ValidateSSHURL rejects all clone URLs except canonical github.com SSH/HTTPS forms.
func ValidateSSHURL(url string) error {
	if url == "" {
		return errors.New("clone URL must not be empty")
	}
	if httpsURLRe.MatchString(url) || sshURLRe.MatchString(url) {
		return nil
	}
	return fmt.Errorf("unexpected clone URL: %q", url)
}

// EnsureClean resets repoDir to HEAD, drops untracked files, removes operation markers.
func EnsureClean(ctx context.Context, repoDir string) error {
	for _, args := range [][]string{
		{"git", "reset", "--hard", "HEAD"},
		{"git", "clean", "-fd"},
	} {
		cmd := exec.CommandContext(ctx, args[0], args[1:]...)
		cmd.Dir = repoDir
		_ = cmd.Run() // mirror Python: check=False
	}
	for _, marker := range []string{"MERGE_HEAD", "CHERRY_PICK_HEAD", "REBASE_HEAD"} {
		_ = os.Remove(filepath.Join(repoDir, ".git", marker))
	}
	return nil
}

// AcquireLock takes an exclusive non-blocking advisory lock on path. Returns
// a release func. Platform-specific implementations live in lock_unix.go and
// lock_windows.go.

// BlastRadius parses git status --porcelain -z and returns sorted touched paths
// excluding README.md. Empty result = clean to ship.
func BlastRadius(ctx context.Context, repoDir string) ([]string, error) {
	cmd := exec.CommandContext(ctx, "git", "status", "--porcelain", "-z")
	cmd.Dir = repoDir
	out, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("git status: %w", err)
	}
	var paths []string
	tokens := strings.Split(string(out), "\x00")
	for i := 0; i < len(tokens); i++ {
		entry := tokens[i]
		if len(entry) < 4 {
			continue
		}
		path := entry[3:]
		// Rename (R) and copy (C) records emit "<XY> <new>\0<old>\0".
		// Consume the trailing old-path token so it doesn't get re-parsed
		// as if it had a status prefix.
		if entry[0] == 'R' || entry[0] == 'C' || entry[1] == 'R' || entry[1] == 'C' {
			if i+1 < len(tokens) {
				old := tokens[i+1]
				i++
				if old != "" && old != "README.md" {
					paths = append(paths, old)
				}
			}
		}
		if path == "README.md" {
			continue
		}
		paths = append(paths, path)
	}
	sort.Strings(paths)
	return paths, nil
}
